torch/_export/graph_module.py

```diff
@@ -224,61 +193,131 @@ def op_to_str(op):
-        inline_constraints = root.meta.get("inline_constraints", {})
-
-    gm = fx.GraphModule(root, graph, class_name)
-
-    input_tracker = 0
-
-    # group by input id
-    input_shape_constraints_by_tensor_id = defaultdict(list)
-    for constraint in input_shape_constraints:
-        input_shape_constraints_by_tensor_id[constraint["t_id"]].append((constraint["dim"], constraint["min"], constraint["max"]))
-
-    input_shape_constraints_by_src_name: Dict[str, List[Tuple[int, ConstraintExpr, ConstraintExpr]]] = {}
-    input_name_to_example_inputs: Dict[str, Any] = {}
-    if example_inputs is not None:
-        input_tracker = 0
-        for node in gm.graph.nodes:
-            if node.op == "placeholder":
-                example_input = example_inputs[input_tracker]
-                if id(example_input) in input_shape_constraints_by_tensor_id:
-                    input_shape_constraints_by_src_name[node.name] = input_shape_constraints_by_tensor_id[id(example_input)]
-                input_name_to_example_inputs[node.name] = example_input
-                input_tracker += 1
-
-    meta = ExportMetadata(
-        in_spec=in_spec,
-        out_spec=out_spec,
-        update_spec=0,
-        mutation=mutation if mutation else [],
-        input_shape_constraints=input_shape_constraints_by_src_name,
-        inline_constraints=inline_constraints,
-        input_name_to_example_inputs=input_name_to_example_inputs,
-    )
-    attach_export_graph_metadata(gm, meta)
+) -> torch.fx.GraphModule:
+    gm = torch.fx.GraphModule(root, graph, class_name=class_name)
+    gm.__class__ = type(type(gm).__name__, (ExportGraphModuleMixin, type(gm)), {})
     return gm
 
 
+class ExportedProgram:
+    def __init__(
+        self,
+        root: Union[torch.nn.Module, Dict[str, Any]],
+        graph: torch.fx.Graph,
+        graph_signature: Optional[GraphSignature] = None,
+        call_spec: Optional[CallSpec] = None,
+        state_dict: Optional[Dict[str, Any]] = None,
+        example_inputs: Any = None,
+    ):
+        # Remove codegen related things from the graph. It should just be a flat graph.
+        graph._codegen = torch.fx.graph.CodeGen()
+        graph_module = make_export_graph_module(root, graph)
+        self.graph_module: ExportGraphModule = graph_module
+
+        self.graph_signature: Optional[GraphSignature] = graph_signature
+        self.call_spec: Optional[CallSpec] = call_spec
+        self.state_dict: Optional[Dict[str, Any]] = state_dict
+        self.symbol_to_range = {}
+
+        input_shape_constraints = (
+            root.meta.get("input_shape_constraints", [])
+            if isinstance(root, torch.fx.GraphModule)
+            else []
+        )
+        self.symbol_to_range = (
+            root.meta.get("inline_constraints", {})
+            if isinstance(root, torch.fx.GraphModule)
+            else {}
+        )
+
+        # group by input id
+        input_shape_constraints_by_tensor_id = defaultdict(list)
```

Comment:
This processing logic is only relevant when we are actually exporting in the first place. Could we make it into a separate preprocessing function that we call when we are creating ExportedProgram? That way, we don't have to keep computing it.

Reply:
Makes sense, I moved it into a private function in __init__.py. Lmk what you think.
cc @avikchaudhuri